lib/adb-device.js

```diff
@@ -190,6 +190,64 @@ class ADBDevice {
     return data;
   }
 
+  // Receive a file listing and parse it into an array of file objects.
+  async recvList(localId, remoteId) {
+    // This is the structure of each directory entry returned by LIST
+    // see https://android.googlesource.com/platform/system/core/+/master/adb/SYNC.TXT#39
+    const props = [
+      'cmd',
+      'mode',
+      'size',
+      'modified',
+      'length',
+      'filename'
+    ];
+    const files = [];
+    let file = {};
+    let i = 0;
+
+    // Keep requesting new packets until we receive 'DONE'
+    // TODO: Do we need a timeout in case 'DONE' is never sent? Shouldn't happen
+    while (true) {
+      let packet = await this.recvAndOkay(localId, remoteId);
+      let pos = 0;
+      do {
```

Comment:
I feel a `do while` structure for parsing the packet a bit more difficult to follow than the parsing maybe needs to be? Maybe there's something I'm missing. From your documentation of the protocol it sounds like the 4 byte fields `mode`, `size`, `modified`, and `length` are always present if the `cmd` is `DENT`? Is that correct. If so can we make this a simple `if else`, read the first 4 bytes and check the command, then pass the remainder of the buffer to a parsing function that just reads each 4 byte field and the filename if the command was `DENT`? What do you think?